src/hc32f460/adc.c

```diff
@@ -0,0 +1,155 @@
+// ADC functions on Huada HC32F460
+//
+// Copyright (C) 2022  Steven Gotthardt <gotthardt@gmail.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include "generic/misc.h"   // timer_from_us
+#include "command.h"        // shutdown
+#include "board/gpio.h"     // gpio_adc_setup
+#include "board/internal.h" // GPIO
+#include "sched.h"          // sched_shutdown
+
+// library
+#include "hc32f460_adc.h"
+#include "hc32f460_pwc.h"
+#include "hc32f460_gpio.h"
+
+#define ADC_RESOLUTION_12BIT        (12u)
+#define ADC_RESOLUTION_10BIT        (10u)
+#define ADC_RESOLUTION_8BIT         (8u)
+
+#define ADC1_RESOLUTION             (ADC_RESOLUTION_12BIT)
+#define ADC1_PRECISION              (1ul << ADC1_RESOLUTION)
+
+#if ADC1_RESOLUTION == ADC_RESOLUTION_12BIT
+#define AdcResolution   AdcResolution_12Bit
+#elif ADC1_RESOLUTION == ADC_RESOLUTION_10BIT
+#define AdcResolution   AdcResolution_10Bit
+#else
+#define AdcResolution   AdcResolution_8Bit
+#endif
+
+
+/* Timeout value definitions. Found in example code */
+#define TIMEOUT_VAL     (30u)
+
+DECL_CONSTANT("ADC_MAX", ADC1_PRECISION-1);
+
+// These pins can be used for ADC
+static const uint8_t adc_gpio[] = {
+    GPIO('A', 0),   // Chan 0
+    GPIO('A', 1),   // Chan 1
+    GPIO('A', 2),   // Chan 2
+    GPIO('A', 3),   // Chan 3
+    GPIO('A', 4),   // Chan 4
+    GPIO('A', 5),   // Chan 5
+    GPIO('A', 6),   // Chan 6
+    GPIO('A', 7),   // Chan 7
+    GPIO('B', 0),   // Chan 8
+    GPIO('B', 1),   // Chan 9
+    GPIO('C', 0),   // Chan 10  // TBed  on TriGorilla
+    GPIO('C', 1),   // Chan 11  // THead on TriGorilla
+    GPIO('C', 2),   // Chan 12
+    GPIO('C', 3),   // Chan 13
+    GPIO('C', 4),   // Chan 14  // TBed  on aquilla
+    GPIO('C', 5),   // Chan 15  // THead on aquilla
+};
+
+
+struct gpio_adc
+gpio_adc_setup(uint32_t gpio)
+{
+    // validate pin in adc_pins table
+    int chan;
+    for (chan=0; ; chan++)
+    {
+        if (chan >= ADC1_CH_COUNT-1)   // don't allow internal
```

Comment:
FWIW, I think the code would be more readable if this did `if (chan >= ARRAY_SIZE(adc_gpio))` as otherwise it's not clear the following access to `adc_gpio` is valid.

Reply:
I incorporated that